File: services/giveaway_service.py
import discord
import asyncio
import random
from datetime import datetime, timezone
from core.database import db
from core.logger import logger
from ui.embeds import SuccessEmbed, AstraEmbed

class GiveawayService:
    async def create_giveaway(self, message_id: int, guild_id: int, channel_id: int, host_id: int, prize: str, winner_count: int, ends_at: datetime):
        await db.execute(
            """
            INSERT INTO giveaways (message_id, guild_id, channel_id, host_id, prize, winner_count, ends_at)
            VALUES (?, ?, ?, ?, ?, ?, ?)
            """,
            message_id, guild_id, channel_id, host_id, prize, winner_count, ends_at
        )

    async def add_entry(self, message_id: int, user_id: int) -> bool:
        """Adds a user to the giveaway. Returns False if already entered."""
        try:
            await db.execute(
                "INSERT INTO giveaway_entries (message_id, user_id) VALUES (?, ?)",
                message_id, user_id
            )
            return True
        except:
            return False

    async def pick_winners(self, message_id: int):
        giveaway = await db.fetch_one("SELECT * FROM giveaways WHERE message_id = ?", message_id)
        if not giveaway or giveaway['is_ended']:
            return []

        entries = await db.fetch_all("SELECT user_id FROM giveaway_entries WHERE message_id = ?", message_id)
        if not entries:
            return []

        user_ids = [row['user_id'] for row in entries]
        winner_count = min(len(user_ids), giveaway['winner_count'])
        
        winners = random.sample(user_ids, winner_count)
        
        await db.execute("UPDATE giveaways SET is_ended = 1 WHERE message_id = ?", message_id)
        return winners

    async def get_active_giveaways(self):
        return await db.fetch_all("SELECT * FROM giveaways WHERE is_ended = 0 AND ends_at <= CURRENT_TIMESTAMP")

giveaway_service = GiveawayService()

File: src/services/giveawayService.ts
import { Client, EmbedBuilder, TextChannel } from 'discord.js';
import { db } from '../core/database';
import { VERSION } from '../core/constants';
import logger from '../core/logger';

export class GiveawayService {
    static async startChecker(client: Client) {
        setInterval(async () => {
            const due = await db.fetchAll(
                "SELECT * FROM giveaways WHERE ended = FALSE AND ends_at <= ?",
                new Date().toISOString()
            );

            for (const giveaway of due) {
                try {
                    const guild = await client.guilds.fetch(giveaway.guild_id).catch(() => null);
                    if (!guild) continue;
                    const channel = await guild.channels.fetch(giveaway.channel_id).catch(() => null) as TextChannel | null;
                    if (!channel) continue;
                    const message = await channel.messages.fetch(giveaway.message_id).catch(() => null);
                    if (!message) continue;

                    const reaction = message.reactions.cache.get('🎉');
                    let mentions = 'No eligible entries.';
                    let winnerIds = '';

                    if (reaction) {
                        const users = await reaction.users.fetch();
                        const eligible = [...users.filter(u => !u.bot).values()];
                        if (eligible.length > 0) {
                            const count = Math.min(giveaway.winners, eligible.length);
                            const selected = eligible.sort(() => Math.random() - 0.5).slice(0, count);
                            mentions = selected.map(w => `<@${w.id}>`).join(', ');
                            winnerIds = selected.map(w => w.id).join(',');
                        }
                    }

                    const embed = new EmbedBuilder()
                        .setColor(0x06D6A0)
                        .setTitle('🎉 Giveaway Ended!')
                        .setDescription(
                            `> **${giveaway.prize}**\n\n` +
                            `🏆 **Winner${giveaway.winners !== 1 ? 's' : ''}:** ${mentions}\n\n` +
                            `Congratulations! 🎊`
                        )
                        .setFooter({ text: `Astra ${VERSION} Nova • Giveaway` })
                        .setTimestamp();

                    await channel.send({ content: winnerIds ? mentions : undefined, embeds: [embed] });
                    await db.execute(
                        'UPDATE giveaways SET ended = TRUE, winner_ids = ? WHERE id = ?',
                        winnerIds, giveaway.id
                    );
                } catch (err) {
                    logger.error(`Giveaway auto-end failed for ID ${giveaway.id}: ${err}`);
                    await db.execute('UPDATE giveaways SET ended = TRUE WHERE id = ?', giveaway.id);
                }
            }
        }, 60000); // Check every 60 seconds
    }
}
